django/thunderstore/permissions/mixins.py

```diff
@@ -41,11 +42,44 @@ class VisibilityMixin(models.Model):
         on_delete=models.PROTECT,
     )
 
+    @transaction.atomic
+    def update_visibility(self):
+        pass
+
     @transaction.atomic
     def save(self, *args, **kwargs):
         if not self.pk and not self.visibility:
-            self.visibility = VisibilityFlags.objects.create_public()
+            self.visibility = VisibilityFlags.objects.create_unpublished()
+
+        self.update_visibility()
+
         super().save()
 
     class Meta:
         abstract = True
+
+    def is_visible_to_user(self, user: UserType) -> bool:
+        if not self.visibility:
+            return False
+
+        if self.visibility.public_detail:
+            return True
+
+        if user is None:
+            return False
+
+        if hasattr(self, "package"):
+            if self.visibility.owner_detail:
+                if self.package.owner.can_user_access(user):
+                    return True
+
+            if self.visibility.moderator_detail:
+                for listing in self.package.community_listings.all():
+                    if listing.community.can_user_manage_packages(user):
+                        return True
+
+        if self.visibility.admin_detail:
+            if user.is_superuser:
+                return True
+
+        return False
```

Comment:
This isn't quite what I was after. The main point was that any location doing visibility checks should only resolve the following 3 booleans based on the context:
- Is the user the owner of the content?
- Is the user a moderator in respect to the content?
- Is the user an admin in respect to the content?

In other words, the function here should mirror the query implementation in terms of the calling convention, e.g.

```python
    def is_visible_detail(self, is_owner: bool, is_moderator: bool, is_admin: bool) -> True:
        return (
            self.visibility.public_detail or
            (self.visibility.owner_detail and is_owner) or
            (self.visibility.moderator_detail and is_moderator) or
            (self.visibility.admin_detail and is_admin)
        )
```

Then on the call site you'd do context-specific resolution of the input attributes before calling the function, e.g.

```python
# The functions are imaginary
is_owner = self.is_user_owner(user)
is_moderator = self.is_user_moderator(user)
is_admin = self.is_user_admin(user)

is_visible = self.is_visible_detail(is_owner, is_moderator, is_admin)
```

The point of the design is to force the codebase to
1. Have to resolve the context-specific visibility roles at the call site of the visibility check
2. Do so in a way that can be efficiently queried against the visibility flags on the DB side (i.e. no complicated business logic involved in the checks themselves; just boolean comparison)

Reply:
99% of the time is_visible_to_user(self, user: UserType) will resolve after
```
        if self.visibility.public_detail:
            return True
```

Won't this be more efficient than always calculating is_user_moderator(user) and is_user_admin(user) every time?

Reply:
So the main benefit of this convention comes from forcing the visibility logic to be resolved in a way that translates to efficient db queries at scale; what happens in a single function that gets run once per request on the python level is trivial by comparison. Even if resolving the flag itself would be slow, querying matching objects from the db will be fast. Right now the paradigm is to fetch an object and then checking if the user has access to the object, whereas we should be able to do both at the same time soon just by applying the visibility filtering on the db query directly.